guidelines/commit_and_pr_text.md

```diff
@@ -0,0 +1,53 @@
+# Motivation
+Our changelogs (reviewed by users/customers) are autogenerated from PR titles and git commit text.  The better this information is, the more we help our customers.  This information also helps reviewers understand changes, patch managers understand risk/importance, and when hunting for regressions, helps us better understand what is/isn’t likely to be a cause.
+
+# PR Metadata
+
+__PR title:__ Bug NNNN: A short sentence with capitalization and no period or the Jira tag for the issue “Feature TEAM-XXX:”
+
+* Try to describe the problem the user sees, rather than the code change you are doing
+  * I.e. “Console does not work when user clicks X” instead of “Fix drawer notification integration with underlying controller”
+* Prefix bug fixes with the bug number, “Bug XXXXX:” 
+  * (capitalize bug, space, number, then colon)
+* Do not let your PR title get truncated/wrapped into the description
+* Additional prefixes like what release the PR applies to are allowed, but they will be stripped from the changelog (the changelog for 4.4 doesn’t need to say “release-4.4” 
+  * Changelog strips anything between square brackets that has no spaces - i.e. [release-4.4] is stripped but [release 4.4] would not be
+    
+
+__PR description:__  Additional context reviewers need to understand your PR
+
+* Why was this change needed?
+* What changes are being made?  
+* Reference to any prior work(other PRs) or follow up work that will be needed.
+
+[Canonical example](https://github.com/openshift/openshift-apiserver/pull/69)
+
+# Git Commit Best practices
+
+## Number of Commits
+
+Strive for a single commit per PR.
+
+Minimal commits per PR make it easier to review the PR, easier to revert the PR if necessary, easier for someone doing “git blame” to follow the set of involved changes, and encourage smaller “chunks of work” (if it’s multiple commits, think about whether it should also be multiple PRs)
```

Comment:
One thing I try to do is make my vendor commits separate from all other work in the PR. I find it incredibly difficult to review 14,000 vendor changes :)

Reply:
covered by https://github.com/openshift/enhancements/pull/311/files#diff-8092445902fc4142135544e8ba663eb1R35 no?

Reply:
> Minimal commits per PR make it easier to review the PR, [...], easier for someone doing “git blame” to follow the set of involved changes, and encourage smaller “chunks of work” (if it’s multiple commits, think about whether it should also be multiple PRs)

When I read this, I can't help but think there are 2 conflicting things here,
1) minimal commits
2) minimizing the number of commits per PR

1) indeed helps with reviews, git blame, ... 2) conflicts with that in that it will not encourage people to split their work meaningfully since the recommendations push to use only 1 commit per PR.
In my experience, the smaller a commit is, the easier it is to review. If you use `git blame` because you want to understand why a change was made, the smaller the commit is, the more specific the commit message will be as to why the change was made. If you use `git bisect` to track down a bug on code you are not much familiar with, you also want to end up on a commit making as little changes as possible. Except for trivial changes/bug fixes, it's usually possible to split a change in several meaning full commits.
Actually, looking at the commit log of the 'canonical example' below ( https://github.com/openshift/origin/commit/d7a7ff790f87ae14ad106f2256ebafc1e7c91949 ), I'm under the impression each paragraph could have deserved its own commit, and this would have made it easier to assess (for example) whether "Remove the 'break' from the wait-loop's ClusterOperator iteration. The previous logic would only list the first unavailable operator. With this commit we now list each unavailable operator while we wait." does not introduce unwanted/unexpected logic changes. As it is now, the change is drowned in other unrelated changes, making the review process harder, and making a potential revert of just this part harder too.

Reply:
I'm changing the text to `Minimize your number of commits per PR with a goal of a single commit in many cases.`

(again with the reasons and exceptions listed below that text).  

> The 1st item indeed helps with reviews, git blame, ... 2) conflicts with that in that it will not encourage people to split their work meaningfully since the recommendations push to use only 1 commit per PR.

Minimal *number* of commits helps with PR review imho, mainly because if you've got a PR with a vendor commit and then 50 other commits, I can't easily exclude the vendor commit from my review (I can either view all changes, or i have to go through each commit by itself).  With a vendor commit plus one or two commits, I can easily review the one or two commits that aren't the vendor commit.  That was the genesis of the statement that "minimal commits per PR make it easier to review the PR".

Certainly to some extent this probably depends on the nature of the PR as well as how reviewers prefer to operate.

Minimal sized commits (which weren't really discussed in this doc, i will rephrase "minimal commits to minimal numbers of commits, in the text) wasn't intended as a goal of this doc.  Commits should be as big or as small as they need to be do to the related/cohesive task they are intended for.  Sometimes that's 1 line in 1 file and sometimes that's 1000 lines in 1000 files (think utility function refactoring).

> 2) conflicts with that in that it will not encourage people to split their work meaningfully since the recommendations push to use only 1 commit per PR.

I think this is where the "consider if it should have been multiple PRs" statement comes in.


> Actually, looking at the commit log of the 'canonical example' below ( openshift/origin@d7a7ff7 ), I'm under the impression each paragraph could have deserved its own commit, 

point taken, the intent was that this was a good example of commit text (explaining what was done), not necessarily a good example of how to break down a body of work into specific commits.  I will make that clear.

Reply:
> > Actually, looking at the commit log of the 'canonical example' below ( [openshift/origin@d7a7ff7](https://github.com/openshift/origin/commit/d7a7ff7) ), I'm under the impression each paragraph could have deserved its own commit,
> 
> point taken, the intent was that this was a good example of commit text (explaining what was done), not necessarily a good example of how to break down a body of work into specific commits. I will make that clear.

For what it's worth, it was clear that it was only a specific example in the context of 'good commit text', it was just convenient to reuse this one as a counter example.


> Minimal _number_ of commits helps with PR review imho, mainly because if you've got a PR with a vendor commit and then 50 other commits, I can't easily exclude the vendor commit from my review (I can either view all changes, or i have to go through each commit by itself). With a vendor commit plus one or two commits, I can easily review the one or two commits that aren't the vendor commit. That was the genesis of the statement that "minimal commits per PR make it easier to review the PR".
> 
> Certainly to some extent this probably depends on the nature of the PR as well as how reviewers prefer to operate.

Hmm it seems we definitely operate very differently, I have a hard time reviewing efficiently/correctly commits doing multiple things which could have been separated in smaller logical commits :-/ But maybe your "50 commits" example would be 50 commits such as "adjust previous commit", "fix bug in 5th commit of PR", "remove debug code added in PR", ...? In which case, yes, I agree it would be complicated to review.

> Minimal sized commits (which weren't really discussed in this doc, i will rephrase "minimal commits to minimal numbers of commits, in the text) wasn't intended as a goal of this doc.

Which I'm worried is going to be a disincentive to splitting commits. As the developer, I'd just read that `git commit -a` and then PR is recommended by this doc :) This imo makes the reviewers task more complicated.

Reply:
> But maybe your "50 commits" example would be 50 commits such as "adjust previous commit", "fix bug in 5th commit of PR", "remove debug code added in PR", ...? In which case, yes, I agree it would be complicated to review.

That is definitely the clearest example of the kind of commit behavior we want to avoid.

> Which I'm worried is going to be a disincentive to splitting commits. As the developer, I'd just read that git commit -a and then PR is recommended by this doc :) This imo makes the reviewers task more complicated.

well again, the idea here is that should be your default mode of operation(and that before you create that second commit, you consider whether it should instead be a second PR), but that there are valid reasons to have multiple commits.